Changed to coreboot patch not to call prog_segment_loaded in smm.

persmule opened the pull request:
If USE_OPTION_TABLE is not set when building coreboot, cbfs_boot_map_with_leak() within src/lib/cbfs.c is not called by smm codes, and will be stripped out at link time, otherwise, other modules may call cbfs_boot_map_with_leak() within smm codes, which breaks its build for prog_segment_loaded() is not built for smm codes (tested when building for Thinkpad x230 with USE_OPTION_TABLE enabled).
I believe that prog_segment_loaded() is not intended to be called within smm codes, so I create this patch, not to call prog_segment_loaded in smm, which enables us to [set USE_OPTION_TABLE and STATIC_OPTION_TABLE to make tuning cmos options easier while keeping security](https://github.com/osresearch/heads/issues/274).

Commit:
Changed to coreboot patch not to call prog_segment_loaded in smm.

flammit commented:
Thanks for the PR.  Seems reasonable, but can you help me with a test config so I can see how this impacts the TPM measurements and test locally?

persmule commented:
What should I do to help you? Have you observed any regression with my commit?

persmule commented:
Try this: [coreboot-x230-optable.config.txt](https://github.com/flammit/heads/files/1440412/coreboot-x230-optable.config.txt)?

flammit commented:
Thanks - I'll check this out later this week when I'm back in front of my x230 test machine.

flammit commented:
Thanks again for the PR and sorry for the delay.  I understand now the point here, but I'm wondering if it is better to remove the cbfs/cmos option dependency in SMM for x230 completely (I don't believe it is used as I've compiled a patch without it).

Also this is an issue for coreboot 4.5 too.  Would you mind moving this to a PR on the main @osresearch master branch so we can get his feedback as well?

persmule commented:
New commit added: afd3a005e078420bbbcfb8194fc90e02dcc25666 , also PRed to [osresearch](https://github.com/osresearch)

flammit commented:
Thanks @persmule - I'm traveling now but will take a look at this over the weekend.  I'll be updating my master to merge coreboot-4.6 soon.  Any thoughts on removing the cbfs dependency on SMM completely?

persmule commented:
> Any thoughts on removing the cbfs dependency on SMM completely?

I do not know, so I tend to make least changes to existing codes to achieve my goal. If you want to proceed removing the cbfs dependency on SMM completely, you had better ask coreboot's developers first, e.g. on #coreboot@irc.freenode.net or their mailing list.

persmule commented:
for [your gist](https://gist.github.com/flammit/d00210803849c9dee66a0cc931152242), I suggest you not make patch for x230 only, but for every board with TPM available or installable (e.g. gigabyte ga-b75m-d3h) coreboot CAN support, so if you touched drivers/pc80/rtc/mc146818rtc.c, please check codes for other RTC chips as well, otherwise you had better make patches for coreboot independent to any specific board. That is why I tend to make least changes to existing codes to achieve my goal.

persmule commented:
The primary part of Heads is a payload based on linux kernel and initrd which in theory can run atop any coreboot build, just like [what I wrote in this article](https://github.com/hardenedlinux/Debian-GNU-Linux-Profiles/blob/master/docs/hardened_boot/heads-atop-coreboot.md).

Commit:
Add OHCI and UHCI drivers to initrd.

USB smart card readers are most full speed devices, and there is no
"rate-matching hubs" beneath the root hub on older (e.g. GM45) plat-
forms, which has companion OHCI or UHCI controllers and needs cor-
responding drivers to communicate with card readers directly plugged
into the motherboard, otherwise a discrete USB hub should be inserted
between the motherboard and the reader.

This time I make inserting linux modules for OHCI and UHCI controllable
with option CONFIG_LINUX_USB_COMPANION_CONTROLLER.

Tested on my x200s and elitebook revolve 810g1.
